py/htm_advanced/algorithms/monitor_mixin/monitor_mixin_base.py

```diff
@@ -0,0 +1,223 @@
+# ----------------------------------------------------------------------
+# HTM Community Edition of NuPIC
+# Copyright (C) 2017, Numenta, Inc. 
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.    If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+MonitorMixinBase class used in monitor mixin framework.
+
+Using a monitor mixin with your algorithm
+-----------------------------------------
+
+1. Create a subclass of your algorithm class, with the first parent being the
+corresponding Monitor class. For example,
+
+        class MonitoredTemporalMemory(TemporalMemoryMonitorMixin,
+                                                                    TemporalMemory): pass
+
+2. Create an instance of the monitored class and use that.
+
+        instance = MonitoredTemporalMemory()
+        # Run data through instance
+
+3. Now you can call the following methods to print monitored data from of your
+instance:
+
+- instance.mmPrettyPrintMetrics(instance.mmGetDefaultMetrics())
+- instance.mmPrettyPrintTraces(instance.mmGetDefaultTraces())
+
+Each specific monitor also has specific methods you can call to extract data
+out of it.
+
+Adding data to a monitor mixin
+-----------------------------------------
+
+1. Create a variable for the data you want to capture in your specific monitor's
+`mmClearHistory` method. For example,
+
+        self._mmTraces["predictedCells"] = IndicesTrace(self, "predicted cells")
```

Comment:
For my own usage, I wanted to get the framework functionally working as close as possible to the original htmresearch implementation so I was not particularly critical of the measurement code as long as everything worked.
I didn't notice the Metrics capability in the bindings. Now that you brought my attention to it, I'll have a look and see what is possible.

Reply:
I don't think this can be replaced with the existing SDR Metrics class.  This code appears to keep a full history of the data, so that it can do a more thorough analysis (plot & histogram).  The existing metrics class does a limited analysis on the data when it is received, but it does not save the data.  This difference is critical: long running applications can not save every bit of data forever, and lab-experiments need to save every bit of data.

On the whole, I think this new metrics class looks useful and I hear that it works.  It's also baked into the research code base, so refactoring it is non-trivial.

Reply:
> On the whole, I think this new metrics class looks useful and I hear that it works. It's also baked into the research code base, so refactoring it is non-trivial.

It definitely works. The example applications use it heavily.

Reply:
Sounds fair, let's keep both metrics as is. 

> I wanted to get the framework functionally working as close as possible to the original htmresearch implementation so I was not particularly critical of 

actually, this is what I'm trying to figure for this PR, be happy that it works and looks as closely as the original repo, or look more at integration with the current code..

Reply:
It is literally a straight Python2 to Python3 port but based on htm.core rather than nupic.core and with no embellishments.
The unit tests are the original unit tests from htmresearch (ported of course)